canisters/src/vts/src/lib.rs

```diff
@@ -47,12 +85,61 @@ fn request_firmware() -> VTSResult<()> {
     Ok(())
 }
 
+#[derive(CandidType, Deserialize, Debug)]
+pub struct UploadFirmwareRequest {
+    pub principal: String,
+    pub _firmware: Vec<u8>,
+    pub _arch: String,
+}
+
 #[ic_cdk::update]
 fn upload_firmware(req: UploadFirmwareRequest) -> VTSResult<()> {
     // todo: this canister method should be executed only by our gateway.
     FIRMWARE_REQUESTS.with(|requests| requests.borrow_mut().remove(&req.principal));
     Ok(())
 }
 
+#[ic_cdk::update]
+fn create_agreement(
+    name: String,
+    vh_customer: Principal,
+    daily_usage_fee: String,
+    gas_price: String,
+) -> VTSResult<u128> {
+    let caller = ic_cdk::api::caller();
+    ic_cdk::println!("requested agreement creation by {}", caller);
+    AGREEMENTS.with(|agreements| {
+        let mut agreements = agreements.borrow_mut();
+
+        if agreements.iter().any(|(_, agreement)| {
```

Comment:
If we cannot understand that new agreement is a duplicate without iterating over all agreements, it is better to delete such logic. Because iterating over all agreements is not possible in smart contract. Let's delete this check, error variant and test.

Reply:
You can test that we can create two agreements for the same entities otherwise :)